File: hermes-kotlin/src/main/kotlin/hermes/search/Fts.kt
package hermes.search

import hermes.KnowledgeGraph
import hermes.ftsSearch

private const val FTS_LIMIT = 20
private const val STRATEGY_MIN_RESULTS = 3
private const val MAX_QUERY_WORDS = 10

/**
 * Returns true for characters that belong to scripts without whitespace word
 * boundaries (CJK ideographs, Hiragana, Katakana, Hangul).
 */
private fun isCjk(ch: Char): Boolean {
    return ch in '\u3040'..'\u309F' ||   // Hiragana
           ch in '\u30A0'..'\u30FF' ||   // Katakana
           ch in '\u3400'..'\u4DBF' ||   // CJK Extension A
           ch in '\u4E00'..'\u9FFF' ||   // CJK Unified Ideographs
           ch in '\uF900'..'\uFAFF' ||   // CJK Compatibility
           ch in '\uAC00'..'\uD7AF'      // Hangul Syllables
}

/**
 * Extracts alphanumeric/underscore tokens from the raw query and removes
 * FTS operators. CJK characters are emitted individually.
 */
fun extractWords(query: String): List<String> {
    val words = mutableListOf<String>()
    val cur = StringBuilder()

    for (ch in query) {
        if (isCjk(ch)) {
            if (cur.isNotEmpty()) {
                if (!isFtsOperator(cur.toString())) words.add(cur.toString())
                cur.clear()
            }
            words.add(ch.toString())
        } else if (ch.isLetterOrDigit() || ch == '_') {
            cur.append(ch)
        } else if (cur.isNotEmpty()) {
            if (!isFtsOperator(cur.toString())) words.add(cur.toString())
            cur.clear()
        }
    }
    if (cur.isNotEmpty() && !isFtsOperator(cur.toString())) {
        words.add(cur.toString())
    }
    return words.take(MAX_QUERY_WORDS)
}

fun ftsSearch(graph: KnowledgeGraph, query: String): List<SearchResult> {
    val words = extractWords(query)
    if (words.isEmpty()) return emptyList()

    if (words.size == 1) {
        val single = "\"${words[0]}\""
        return toSearchResults(graph.ftsSearch(single, FTS_LIMIT))
    }

    // Strategy 1: phrase match
    val phraseQuery = "\"${words.joinToString(" ")}\""
    val s1 = graph.ftsSearch(phraseQuery, FTS_LIMIT)
    if (s1.size >= STRATEGY_MIN_RESULTS) return toSearchResults(s1)

    // Strategy 2: AND with prefix
    val andQuery = words.joinToString(" AND ") { "\"$it\"*" }
    val s2 = graph.ftsSearch(andQuery, FTS_LIMIT)
    if (s2.size >= STRATEGY_MIN_RESULTS) return toSearchResults(s2)

    // Strategy 3: OR
    val orQuery = words.joinToString(" OR ") { "\"$it\"" }
    return toSearchResults(graph.ftsSearch(orQuery, FTS_LIMIT))
}

private fun toSearchResults(raw: List<Pair<hermes.Node, Double>>): List<SearchResult> {
    return raw.map { (node, rank) ->
        SearchResult(
            node = node,
            score = normalizeBm25Score(rank),
            tier = SearchTier.L1_FTS,
            matchedContent = null
        )
    }
}

private fun isFtsOperator(word: String): Boolean {
    return word.uppercase() in listOf("AND", "OR", "NOT", "NEAR")
}

fun normalizeBm25Score(rank: Double): Double {
    val absRank = kotlin.math.abs(rank)
    if (absRank < 0.001) return 0.5
    return (1.0 - 1.0 / (1.0 + absRank)).coerceAtMost(1.0)
}

File: tools/hermes-engine/src/search/fts.rs
// ChartApp/hermes-engine/src/search/fts.rs
use crate::graph::{KnowledgeGraph, Node};
use crate::search::{SearchResult, SearchTier};
use anyhow::Result;

const FTS_LIMIT: usize = 20;
const STRATEGY_MIN_RESULTS: usize = 3;
const MAX_QUERY_WORDS: usize = 10;

/// Task 2.1: Three-strategy FTS with phrase → AND-prefix → OR fallback.
/// Truncates to first 10 meaningful words to avoid degenerate queries on long strings.
pub fn fts_search(graph: &KnowledgeGraph, query: &str) -> Result<Vec<SearchResult>> {
    let words = extract_query_terms(query);

    if words.is_empty() {
        return Ok(Vec::new());
    }

    if words.len() == 1 {
        let single = quote_fts_term(&words[0]);
        return Ok(to_search_results(graph.fts_search(&single, FTS_LIMIT)?));
    }

    // Strategy 1: Exact phrase match — highest precision
    let phrase_query = format!("\"{}\"", words.join(" "));
    let s1 = graph.fts_search(&phrase_query, FTS_LIMIT)?;
    if s1.len() >= STRATEGY_MIN_RESULTS {
        return Ok(to_search_results(s1));
    }

    // Strategy 2: AND-prefix match — good recall for multi-token queries
    let and_query = words
        .iter()
        .map(|w| format!("{}*", quote_fts_term(w)))
        .collect::<Vec<_>>()
        .join(" AND ");
    let s2 = graph.fts_search(&and_query, FTS_LIMIT)?;
    if s2.len() >= STRATEGY_MIN_RESULTS {
        return Ok(to_search_results(s2));
    }

    // Strategy 3: OR fallback — maximum recall
    let or_query = words
        .iter()
        .map(|w| quote_fts_term(w))
        .collect::<Vec<_>>()
        .join(" OR ");
    Ok(to_search_results(graph.fts_search(&or_query, FTS_LIMIT)?))
}

fn extract_query_terms(query: &str) -> Vec<String> {
    query
        .split_whitespace()
        .filter(|w| !is_fts_operator(w))
        .map(sanitize_term)
        .filter(|w| !w.is_empty())
        .take(MAX_QUERY_WORDS)
        .collect()
}

fn sanitize_term(word: &str) -> String {
    word
        .trim_matches('"')
        .replace('"', "")
        .trim()
        .to_string()
}

fn quote_fts_term(term: &str) -> String {
    format!("\"{}\"", term)
}

fn to_search_results(raw: Vec<(Node, f64)>) -> Vec<SearchResult> {
    raw.into_iter()
        .map(|(node, rank)| SearchResult {
            node,
            score: normalize_bm25_score(rank),
            tier: SearchTier::L1Fts,
            matched_content: None,
        })
        .collect()
}

fn is_fts_operator(word: &str) -> bool {
    matches!(word.to_uppercase().as_str(), "AND" | "OR" | "NOT" | "NEAR")
}

fn normalize_bm25_score(rank: f64) -> f64 {
    let abs_rank = rank.abs();
    if abs_rank < 0.001 {
        return 0.5;
    }
    (1.0 - 1.0 / (1.0 + abs_rank)).min(1.0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::HermesEngine;

    fn make_graph(engine: &HermesEngine) -> crate::graph::KnowledgeGraph {
        crate::graph::KnowledgeGraph::new(engine.db().clone(), engine.project_id())
    }

    #[test]
    fn single_word_uses_phrase_strategy() {
        let sanitized = prepare_test_query("main");
        assert_eq!(sanitized, "\"main\"");
    }

    fn prepare_test_query(query: &str) -> String {
        let words = extract_query_terms(query);
        if words.len() == 1 {
            quote_fts_term(&words[0])
        } else {
            format!("\"{}\"", words.join(" "))
        }
    }

    #[test]
    fn filters_fts_operators() {
        let words = extract_query_terms("NOT main AND test OR foo");
        assert!(!words.iter().any(|w| w == "NOT"));
        assert!(!words.iter().any(|w| w == "AND"));
        assert!(!words.iter().any(|w| w == "OR"));
        assert!(words.iter().any(|w| w == "main"));
        assert!(words.iter().any(|w| w == "test"));
        assert!(words.iter().any(|w| w == "foo"));
    }

    #[test]
    fn truncates_to_ten_words() {
        let long_query = "a b c d e f g h i j k l m n";
        let words = extract_query_terms(long_query);
        assert_eq!(words.len(), MAX_QUERY_WORDS);
    }

    #[test]
    fn strips_user_quotes_from_path_terms() {
        let words = extract_query_terms("\"/api/alerts\" handler");
        assert_eq!(words, vec!["/api/alerts".to_string(), "handler".to_string()]);
    }

    #[test]
    fn single_quoted_path_produces_valid_fts_term() {
        let query = prepare_test_query("\"/api/alerts\"");
        assert_eq!(query, "\"/api/alerts\"");
    }

    #[test]
    fn bm25_normalization() {
        assert!(normalize_bm25_score(-5.0) > 0.5);
        assert!(normalize_bm25_score(-10.0) > normalize_bm25_score(-5.0));
        assert!(normalize_bm25_score(0.0) < 0.6);
    }

    #[test]
    fn empty_query_returns_empty() {
        let engine = HermesEngine::in_memory("test-fts").unwrap();
        let graph = make_graph(&engine);
        let results = fts_search(&graph, "").unwrap();
        assert!(results.is_empty());
    }

    #[test]
    fn operator_only_query_returns_empty() {
        let engine = HermesEngine::in_memory("test-fts").unwrap();
        let graph = make_graph(&engine);
        let results = fts_search(&graph, "AND OR NOT").unwrap();
        assert!(results.is_empty());
    }
}

